### PR TITLE
TASK: Migrate Neos.Fusion/Core to phpstan level 8 (Neos 8.3)

### PR DESCRIPTION
**Pull request by mhsdesign**
Requires: https://github.com/neos/neos-development-collection/pull/4842

Only files in `Neos.Fusion/Classes/Core` have been migrated, as they contain relatively new code like the parser.

Two kind of errors have been partially ignored from the level 8

For one: "no value type specified in iterable type array" as adding `array<int|string, mixed>` would just bloat this commit up. In these files we are mostly dealing with the fusion configuration which is highly dynamic:

- Neos.Fusion/Classes/Core/Runtime.php
- Neos.Fusion/Classes/Core/Cache/RuntimeContentCache.php
- Neos.Fusion/Classes/Core/Cache/ContentCache.php
- Neos.Fusion/Classes/Core/Cache/CacheSegmentParser.php
- Neos.Fusion/Classes/Core/RuntimeConfiguration.php
- Neos.Fusion/Classes/Core/ObjectTreeParser/MergedArrayTreeVisitor.php
- Neos.Fusion/Classes/Core/ObjectTreeParser/MergedArrayTree.php

And "has no return type specified" was noticed a lot in the ast visiting code. But i plan to refactor this instead to something better type- and readable in general:

- Neos.Fusion/Classes/Core/ObjectTreeParser/MergedArrayTreeVisitor.php
- Neos.Fusion/Classes/Core/ObjectTreeParser/AstNodeVisitorInterface.php
- Neos.Fusion/Classes/Core/ObjectTreeParser/Ast/*

**The phpstan config i used for local development & testing:**

<details><summary>local phpstan config</summary>
<p>

```
parameters:
    level: 8
    ignoreErrors:
        -
          message: '#no value type specified in iterable type array.$#'
          paths:
            - Neos.Fusion/Classes/Core/Runtime.php
            - Neos.Fusion/Classes/Core/Cache/RuntimeContentCache.php
            - Neos.Fusion/Classes/Core/Cache/ContentCache.php
            - Neos.Fusion/Classes/Core/Cache/CacheSegmentParser.php
            - Neos.Fusion/Classes/Core/RuntimeConfiguration.php
            - Neos.Fusion/Classes/Core/ObjectTreeParser/MergedArrayTreeVisitor.php
            - Neos.Fusion/Classes/Core/ObjectTreeParser/MergedArrayTree.php
        -
          message: '#has no return type specified.$#'
          paths:
            - Neos.Fusion/Classes/Core/ObjectTreeParser/MergedArrayTreeVisitor.php
            - Neos.Fusion/Classes/Core/ObjectTreeParser/AstNodeVisitorInterface.php
            - Neos.Fusion/Classes/Core/ObjectTreeParser/Ast
    paths:
        - Neos.Fusion/Classes/Core
    bootstrapFiles:
         - ../Framework/bootstrap-phpstan.php
```

</p>
</details> 

As phpstan [doesnt](https://github.com/phpstan/phpstan/discussions/10459) allow to configure multiple levels by path, and i dont want to have to burden running phpstan multiple times, i decided against actually linting the migrated `Neos.Fusion/Core` code in ci for Neos 8.3. On the neos 9 branch this we will activate the full linting power though!

### CODE REVIEW

**Comment by mhsdesign** [comment]
depends on https://github.com/neos/flow-development-collection/pull/3274

File: Neos.Fusion/Classes/Core/ObjectTreeParser/Token.php [hunk 1]
```diff
@@ -89,16 +89,17 @@ public static function typeToString(int $type): string
     {
         $stringRepresentation = array_search($type, static::getConstants(), true);
 
-        if ($stringRepresentation === false) {
+        if (is_string($stringRepresentation) === false) {
```
Comment (dlubitz) [inline Neos.Fusion/Classes/Core/ObjectTreeParser/Token.php:92]: Why this? array_search returns false, in case it doesn't found the value.

**Review by dlubitz — COMMENTED** [review]
Looks good in general. But you've changed a lot of method signatures (return types and argument types). I'm not sure how to handle this, as this might be a BC, right? But on the other hand I assume they are internal. 🤷‍♂️

File: Neos.Fusion/Classes/Core/ObjectTreeParser/Token.php (hunk 1 above)
Comment (mhsdesign) [reply Neos.Fusion/Classes/Core/ObjectTreeParser/Token.php:92]: it returns `string|int|false` but so we have to tell phpstan that its not `int` as well. Thus the different check

**Comment by mhsdesign** [comment]
Yes i made sure to not add return types anywhere where people even remotely would use them. Everything in `ObjectTreeParser` contains the "new" parser and is internal, and not usable really anyhow from the outside.

**Review by dlubitz — APPROVED** [review]

**Review by kitsunet — COMMENTED** [review]
Yeah seems fine to me, docblocks I don't care, and signatures are very few. I guess the main risk would be someone extending the exception handlers and stumbling over the void return, but I don't even see that

**Review by kitsunet — APPROVED** [review]

**Comment by mhsdesign** [comment]
> the exception handlers and stumbling over the void return

Thanks. Yes the fact that the return type of `injectThrowableStorage` is void should not influence anyone ^^

**Comment by mhsdesign** [comment]
Fyi 116 errors left when linting the whole fusion package. Which will be ultimately the goal, but that can be done at another time ;)

**Comment by mhsdesign** [comment]
Followup on 9.0 to fix ci: https://github.com/neos/neos-development-collection/pull/4870

Thanks for your reviews ❤️

**Pull request merged**